File: packages/contracts-ts/src/ohlcv.ts
import { z } from "zod";
import { BarInterval, PITFields } from "./common";

/**
 * OHLCV mumu. Fiyatlar API'de Decimal → JSON'da string gelebilir; z.coerce.number
 * ile güvenle sayıya çevrilir (grafik tüketimi için).
 */
export const OHLCVBar = PITFields.extend({
  tenant_id: z.string().default("default"),
  symbol: z.string(),
  exchange_code: z.string().default("XIST"),
  interval: BarInterval.default("1d"),
  open: z.coerce.number(),
  high: z.coerce.number(),
  low: z.coerce.number(),
  close: z.coerce.number(),
  volume: z.coerce.number(),
  adjusted_close: z.coerce.number().nullable().optional(),
  source: z.string(),
});
export type OHLCVBar = z.infer<typeof OHLCVBar>;

export const OHLCVSeries = z.object({
  symbol: z.string(),
  interval: BarInterval,
  bars: z.array(OHLCVBar),
});
export type OHLCVSeries = z.infer<typeof OHLCVSeries>;

File: services/api/heimdall_api/routers/ohlcv.py
"""OHLCV (piyasa verisi) uçları."""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import text
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from heimdall_contracts import BarInterval, OHLCVBar

from ..db import get_session

router = APIRouter(prefix="/instruments", tags=["market"])

_SELECT = text(
    """
    SELECT tenant_id, symbol, exchange_code, interval, event_time, knowledge_time,
           open, high, low, close, volume, adjusted_close, source
    FROM ohlcv_bar
    WHERE symbol = :symbol AND interval = :interval
    ORDER BY event_time ASC
    LIMIT :limit
    """
)


@router.get("/{symbol}/ohlcv", summary="Bir sembolün OHLCV serisini getir")
def get_ohlcv(
    symbol: str,
    interval: BarInterval = Query(default=BarInterval.D1),
    limit: int = Query(default=500, ge=1, le=5000),
    session: Session = Depends(get_session),
) -> dict[str, Any]:
    params = {"symbol": symbol.upper(), "interval": interval.value, "limit": limit}
    try:
        rows = session.execute(_SELECT, params).mappings().all()
    except SQLAlchemyError as exc:
        raise HTTPException(
            status_code=503,
            detail="Veritabanına ulaşılamadı (Docker/Postgres ayakta mı?).",
        ) from exc

    bars = [OHLCVBar(**dict(row)) for row in rows]
    return {
        "symbol": symbol.upper(),
        "interval": interval.value,
        "bars": [b.model_dump(mode="json") for b in bars],
    }
